### PR TITLE
[SystemZ] Don't lower float/double ATOMIC_[LOAD|STORE] to [LOAD|STORE]

### PR DESCRIPTION
**Pull request by JonPsson1**
Instead of lowering float/double ISD::ATOMIC_LOAD / ISD::ATOMIC_STORE nodes to regular LOAD/STORE nodes,  make them legal and do this transformation in Select() instead.

These nodes should not be exposed to DAGCombiner as they were with the non-atomic opcodes, and the general intent is that atomic operations should have the ATOMIC opcodes.

AtomicExpand pass no longer casts float/double load/stores to integer, but the fp128 still is as there is a special handling needed for 128-bit accesses.

### CODE REVIEW

File: llvm/test/CodeGen/SystemZ/atomic-store-06.ll [hunk 1]
```diff
@@ -1,12 +1,11 @@
...
 ;
 ; RUN: llc < %s -mtriple=s390x-linux-gnu | FileCheck %s
 
 define void @f1(ptr %src, float %val) {
 ; CHECK-LABEL: f1:
-; CHECK: lgdr [[R:%r[0-9]+]], %f0
-; CHECK: srlg [[R]], [[R]], 32
-; CHECK: st [[R]], 0(%r2)
+; CHECK: ste %f0, 0(%r2)
+; CHECK: bcr 15, %r0
```
Comment (uweigand) [inline llvm/test/CodeGen/SystemZ/atomic-store-06.ll:8]: Was the brc 15 missing before, or was just the check missing?  I see the check is also missing in the double version of this test.  These test should probably best be fully auto-generated.

Also, why do the long double versions of the tests use library calls?  Shouldn't they use lpq/stpq plus GPR/FPR moves?   In addition, for the long double (and also the i128) versions, we really should test both z10 and z13.

**Review by uweigand — COMMENTED** [review]
Looks generally good to me, but see a couple of questions inline with the test case.

**Comment by jyknight** [comment]
I don't know if this patch also works, but why not translate atomic_load_8/etc patterns directly in tablegen to instructions, and skip the generic LOAD entirely, like e.g. X86 does:
```
def : Pat<(i8  (atomic_load_8 addr:$src)),  (MOV8rm addr:$src)>;
def : Pat<(i16 (atomic_load_16 addr:$src)), (MOV16rm addr:$src)>;
def : Pat<(i32 (atomic_load_32 addr:$src)), (MOV32rm addr:$src)>;
def : Pat<(i64 (atomic_load_64 addr:$src)), (MOV64rm addr:$src)>;
[...]
def : Pat<(atomic_store_8 (i8 imm:$src), addr:$dst),
          (MOV8mi addr:$dst, imm:$src)>;
def : Pat<(atomic_store_16 (i16 imm:$src), addr:$dst),
          (MOV16mi addr:$dst, imm:$src)>;
def : Pat<(atomic_store_32 (i32 imm:$src), addr:$dst),
          (MOV32mi addr:$dst, imm:$src)>;
def : Pat<(atomic_store_64 (i64immSExt32:$src), addr:$dst),
          (MOV64mi32 addr:$dst, i64immSExt32:$src)>;
```

You only need to deal with the trailing fence after the store; for that, could have a Custom action which, if the store is SequentiallyConsistent, converts it to a Release atomic_store plus the fence, otherwise passes it along.

File: llvm/test/CodeGen/SystemZ/atomic-store-06.ll (hunk 1 above)
Comment (JonPsson1) [reply llvm/test/CodeGen/SystemZ/atomic-store-06.ll:8]: Yes, the CHECK for bcr 15 was missing before - sorry should have mentioned that.

For the long double tests, see https://github.com/llvm/llvm-project/pull/75630#issuecomment-1861290278. I think the way the tests were written the output would actually be the same for z10/z13...

**Comment by JonPsson1** [comment]
> I don't know if this patch also works, but why not translate atomic_load_8/etc patterns directly in tablegen to instructions, and skip the generic LOAD entirely, like e.g. X86 does:
> 

On SystemZ, the memory instructions are only atomic if naturally aligned, so the non-atomic ISD opcode would still be needed for unaligned accesses.

I however now see a problem with this patch as the reg/mem folding does not seem to be working as I had expected, which need to look into.

**Comment by JonPsson1** [comment]
As @arsenm suggested *getting rid* of ATOMIC_LOAD, while I thought the general idea was to clean up LOAD/STORE nodes handling by freeing them of possibly being atomic, I wonder if there is a consensus about this to follow...?

The reason I am asking is that it doesn't seem quite trivial after all to make sure all the reg/mem folding work both for LOADs and ATOMIC_LOADs in the SystemZ backend.

**Comment by jyknight** [comment]
I don't really know what arsenm or @preames has/had in mind, but I think there is value in all the backends working in a similar way here, rather than SystemZ doing something different.

The reason I'm skeptical about merging the two operations is that there is a significant difference between LOAD and ATOMIC_LOAD beyond just the memory-ordering: a LOAD (even in some cases when it's volatile!) can be split (both by SelectionDAG expansions, and potentially within the hardware instruction that was selected). An ATOMIC_LOAD cannot be. Given the difference, I'm not sure merging the two into one opcode actually makes things simpler.

**Comment by JonPsson1** [comment]
I am not sure which I think would be best: it seems natural to simply have "LOAD" as it has the MMO with the atomic bits. On the other hand, it kind of feel safer to have the ATOMIC_LOAD opcode to keep it separate from all optimizations on LOADs, even though guarding that with a simple isAtomic() check might be simple enough.

I guess the next step might be to rework the SystemZ implementation with this patch and even after that the ATOMIC_LOAD could be eliminated at some future point if needed. 

Strictly speaking though, from a SystemZ backend perspective it would be simpler to just have the LOAD opcode as the same instruction is to be selected. Not sure how much work that would be, but we would have to be sure that the MMO is properly preserved everywhere, that optimizations are guarded with an isAtomic() check, and that backends like PowerPC selecting atomic loads with ATOMIC_LOAD learned how to do that by checking the MMO of a LOAD instead.

**Comment by preames** [comment]
> I don't really know what arsenm or @preames has/had in mind,

I was only getting the code back into the state it was in before I'd added complexity which never became worthwhile.  I have no strong opinions on how SystemZ should or should not implement this.  Having atomic loads represented by a normal load with the appropriate MMO isn't a bad idea, it just requires a lot more work through generic DAG than I had time to ever finish.  If SystemZ is lowering late enough to avoid those problems (in practice..), then so be it.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp [hunk 2]
```diff
@@ -338,6 +338,19 @@ SDValue DAGTypeLegalizer::PromoteIntRes_Atomic0(AtomicSDNode *N) {
                               N->getMemoryVT(), ResVT,
                               N->getChain(), N->getBasePtr(),
                               N->getMemOperand());
+  if (N->getOpcode() == ISD::ATOMIC_LOAD) {
+    ISD::LoadExtType ETy = cast<AtomicSDNode>(N)->getExtensionType();
+    if (ETy == ISD::NON_EXTLOAD) {
+      if (TLI.getExtendForAtomicOps() == ISD::SIGN_EXTEND)
+        ETy = ISD::SEXTLOAD;
+      else if (TLI.getExtendForAtomicOps() == ISD::ZERO_EXTEND)
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:346]: switch over TLI.getExtendForAtomicOps() (terrible hook btw, one of the plusses of globalisel having G_SEXTLOAD/G_ZEXTLOAD + atomic MMO)

File: llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp [hunk 3]
```diff
@@ -1545,6 +1552,37 @@ bool SystemZDAGToDAGISel::storeLoadIsAligned(SDNode *N) const {
   return true;
 }
 
+// This is a hack to convert ATOMIC_LOADs to LOADs in the last minute just
```
Comment (arsenm) [inline llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp:1555]: It's also not difficult to use PatFrags between atomic and non-atomic loads instead of hacking this up in C++

Comment (JonPsson1) [reply llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp:1552]: Well, with all the different extending load operators it is not quite simple I'd say, but of course better. My concern is that if the ATOMIC_LOAD opcode will go away those changes would have to be reverted again in the .td files. So if that might happen in some foreseeable future, maybe this hack would be acceptable? As an alternative, maybe the switch to PatFrags could be done as a follow-up patch so that it could more easily be reverted if needed..?

**Comment by JonPsson1** [comment]
Patch updated:

- Don't return false for atomic memops in storeLoadIsAligned().
- Treat ATOMIC_LOADs the same as regular LOADs in isVectorElementLoad().
- The TernaryRXF instructions are atomic - remove the nonatomic_ld operator.

File: llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp (hunk 3 above)
Comment (arsenm) [reply llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp:1552]: I still think getting rid of ATOMIC_LOAD is the correct way to do this, and anything else is working around an aversion to completing the work to get there

**Review by arsenm — APPROVED** [review]
The generic bits lgtm. The systemZ handling is a hack, and we'd still be better off getting rid of ATOMIC_LOAD and following how GlobalISel does this

**Comment by JonPsson1** [comment]
> The generic bits lgtm. The systemZ handling is a hack, and we'd still be better off getting rid of ATOMIC_LOAD and following how GlobalISel does this

Thanks for review! Before committing, I think I will try to get rid of the hack in SystemZ and use PatFrags there instead. I will also make a note somewhere that ATOMIC_LOAD might removed at some point.

File: llvm/lib/Target/PowerPC/PPCInstrInfo.td [hunk 4]
```diff
@@ -5035,12 +5035,12 @@ defm : TrapExtendedMnemonic<"lng", 6>;
...
-def : Pat<(atomic_load_32 DForm:$src), (LWZ  memri:$src)>;
-def : Pat<(atomic_load_8  XForm:$src), (LBZX memrr:$src)>;
-def : Pat<(atomic_load_16 XForm:$src), (LHZX memrr:$src)>;
-def : Pat<(atomic_load_32 XForm:$src), (LWZX memrr:$src)>;
+def : Pat<(iAny (atomic_load_8  DForm:$src)), (LBZ  memri:$src)>;
+def : Pat<(iAny (atomic_load_16 DForm:$src)), (LHZ  memri:$src)>;
+def : Pat<(iAny (atomic_load_32 DForm:$src)), (LWZ  memri:$src)>;
+def : Pat<(iAny (atomic_load_8  XForm:$src)), (LBZX memrr:$src)>;
+def : Pat<(iAny (atomic_load_16 XForm:$src)), (LHZX memrr:$src)>;
+def : Pat<(iAny (atomic_load_32 XForm:$src)), (LWZX memrr:$src)>;
```
Comment (arsenm) [inline llvm/lib/Target/PowerPC/PPCInstrInfo.td:5043]: I would assume these should all be i32?

File: llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp [hunk 5]
```diff
@@ -1742,7 +1758,44 @@ void SystemZDAGToDAGISel::Select(SDNode *Node) {
...
   }
+
+  case ISD::ATOMIC_STORE: {
+    auto *AtomOp = cast<AtomicSDNode>(Node);
+    // Store FP values directly without first moving to a GPR. This is needed
+    // as long as clang always emits the cast to integer.
+    EVT SVT = AtomOp->getMemoryVT();
+    SDValue StoredVal = AtomOp->getVal();
+    if (SVT.isInteger() && StoredVal->getOpcode() == ISD::BITCAST &&
+        StoredVal->getOperand(0).getValueType().isFloatingPoint()) {
```
Comment (arsenm) [inline llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp:1769]: can you instead just make the FP typed atomic_store legal? I would assume the cast + load/store get folded together

**Comment by arsenm** [comment]
> We could do something similar with the atomic_store:s, like z_store, but it seems the handling of stores is quite alright like this, or?

Stores should be similarly handled, you can also use PatFrags

File: llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp (hunk 5 above)
Comment (JonPsson1) [reply llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp:1769]: I think this can go away after #83446 lands.

**Comment by JonPsson1** [comment]
> > We could do something similar with the atomic_store:s, like z_store, but it seems the handling of stores is quite alright like this, or?
> 
> Stores should be similarly handled, you can also use PatFrags

Ideally, yes. This is however much less of a hack than what the handling of loads was before since the store gets selected after its operands and therefore has everything ready.

File: llvm/lib/Target/PowerPC/PPCInstrInfo.td (hunk 4 above)
Comment (JonPsson1) [reply llvm/lib/Target/PowerPC/PPCInstrInfo.td]: thanks - fixed.

**Comment by JonPsson1** [comment]
Patch updated: PPC integer types in patterns specified properly.

Waiting on #83446 which will remove the FP/Int casting that Clang is currently doing.

File: llvm/lib/Target/SystemZ/SystemZISelLowering.cpp [hunk 6]
```diff
@@ -920,6 +916,23 @@ bool SystemZTargetLowering::hasInlineStackProbe(const MachineFunction &MF) const
   return false;
 }
 
+// FIXME: Clang emits these casts always regardless of these hooks.
+TargetLowering::AtomicExpansionKind
+SystemZTargetLowering::shouldCastAtomicLoadInIR(LoadInst *LI) const {
```
Comment (arsenm) [inline llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:921]: Looking at this again, I don't think this target hook should exist. Fundamentally AtomicExpandPass exists as a hack because SelectionDAG cannot handle emitting control flow for the cmpxchg blocks. Bitcasting types to match instructions is best done as part of regular legalization. cc @shiltian

**Comment by JonPsson1** [comment]
Patch updated.

@uweigand Now that the Clang FE does not insert the fp<->int bitcasts, the handling for the resulting bitcasts have been removed from the SystemZ backend (there doesn't seem to be any value at all to having them around on SPEC). @arsenm approved the common code changes earlier, so now it's the SystemZ changes that also need review.

@kaz7 This patch shouldn't affect the VE backend at all, but since SDTAtomicLoad has been changed to also allow FP loads, I had to insert the equivalent MVT specifier in the VE patterns for atomic loads. I used 'iAny' for now, but this could easily be changed.

File: llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp [hunk 7]
```diff
@@ -1742,7 +1758,35 @@ void SystemZDAGToDAGISel::Select(SDNode *Node) {
...
+    if (AtomOp->getSuccessOrdering() == AtomicOrdering::SequentiallyConsistent)
+      Chain = CurDAG->getMachineNode(SystemZ::Serialize, SDLoc(AtomOp),
+                                     MVT::Other, SDValue(Chain, 0));
+    ReplaceNode(Node, Chain);
+    SelectCode(St);
+    return;
   }
+  }
+
+#ifndef NDEBUG
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp:1783]: I don't think we need that - nodes that are invalid that way already should just not be matched, right?

File: llvm/lib/Target/SystemZ/SystemZISelLowering.cpp [hunk 8]
```diff
@@ -6587,6 +6574,27 @@ SDValue SystemZTargetLowering::combineTruncateExtract(
   return SDValue();
 }
 
+// Replace ALoad with a new ATOMIC_LOAD with a result that is extended to VT
+// per ETy.
+static SDValue extendAtomicLoad(AtomicSDNode *ALoad, EVT VT, SelectionDAG &DAG,
+                                ISD::LoadExtType ETy) {
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:6580]: Given that this patch adds extending atomic loads as a common-code feature, shouldn't now also common code attempt to match them in DAGCombine?  Why does this need to be platform-specific?   But I guess this can be looked at in a separate patch.

**Review by uweigand — APPROVED** [review]
See inline comments, but generally this LGTM now.   As Matt suggested, I guess we should also have a `z_store` PatFrag so we keep the ATOMIC_STORE during SelectionDAG.  But this can also be looked at separately, I think.

File: llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp (hunk 7 above)
Comment (JonPsson1) [reply llvm/lib/Target/SystemZ/SystemZISelDAGToDAG.cpp:1783]: ok, removed.

**Comment by JonPsson1** [comment]
Removed check for extending atomic fp load nodes.

> Given that this patch adds extending atomic loads as a common-code feature, shouldn't now also common code attempt to match them in DAGCombine? Why does this need to be platform-specific? But I guess this can be looked at in a separate patch.

ok, will do a follow-up on moving that into common-code.
